### PR TITLE
[InstCombine] Fold an unsigned comparison of `add nsw X, C` with a constant into a signed comparison

### PR DESCRIPTION
**Pull request by Poseydon42**
Given an unsigned integer comparison of `add nsw X, C1` with some constant `C2` we can fold it into a signed comparison of `X` and `C2 - C1` under the following conditions:
  * There's a `nsw` flag on the addition
  * `C2` is non-negative
  * `X + C1` is non-negative
  * `C2 - C1` is non-negative

Proofs: https://alive2.llvm.org/ce/z/EhCw3E

This would help us with a regression that is currently blocking #101049 from being merged.

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 1]
```diff
@@ -3077,6 +3077,13 @@ Instruction *InstCombinerImpl::foldICmpAddConstant(ICmpInst &Cmp,
       return new ICmpInst(Pred, X, ConstantInt::get(Ty, NewC));
   }
 
+  ConstantRange LHS_CR = computeConstantRange(X, true).add(*C2);
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:3080]:
```suggestion
  ConstantRange LHS_CR = computeConstantRange(X, /*ForSigned=*/true).add(*C2);
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 2]
```diff
@@ -3077,6 +3077,13 @@ Instruction *InstCombinerImpl::foldICmpAddConstant(ICmpInst &Cmp,
...
+  if (ICmpInst::isUnsigned(Pred) && Add->hasNoSignedWrap() &&
+      C.isNonNegative() && LHS_CR.isAllNonNegative() &&
+      (C - *C2).isNonNegative())
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:3083]:
```suggestion
      C.isNonNegative() &&
      (C - *C2).isNonNegative() && computeConstantRange(X, /*ForSigned=*/true).add(*C2).isAllNonNegative())
```

`computeConstantRange` is expensive.

File: llvm/test/Transforms/InstCombine/icmp-add.ll [hunk 3]
```diff
@@ -3102,3 +3102,72 @@ define i1 @uge_add_C2_pow2_C_neg(i8 %x) {
...
+define i1 @icmp_add_constant_with_constant_ult_to_slt_neg4(i32 range(i32 -4, 10) %x) {
+; CHECK-LABEL: @icmp_add_constant_with_constant_ult_to_slt_neg4(
+; CHECK-NEXT:    [[ADD:%.*]] = add nsw i32 [[X:%.*]], 5
+; CHECK-NEXT:    [[CMP:%.*]] = icmp ult i32 [[ADD]], 2
+; CHECK-NEXT:    ret i1 [[CMP]]
+;
+  %add = add nsw i32 %x, 5
+  %cmp = icmp ult i32 %add, 2
+  ret i1 %cmp
+}
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/icmp-add.ll:3173]: I'd suggest to also add one test with the original ucmp/scmp pattern, to show that the whole pattern folds down to icmp now.

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/test/Transforms/InstCombine/icmp-add.ll (hunk 3 above)
Comment (nikic) [reply llvm/test/Transforms/InstCombine/icmp-add.ll:3173]: I mean the `(ucmp + 1) ult 2` pattern here.

Comment (Poseydon42) [reply llvm/test/Transforms/InstCombine/icmp-add.ll:3173]: Thanks, added it

**Review by nikic — APPROVED** [review]

**Review by dtcxzyw — APPROVED** [review]
Nice catch!

**Pull request merged**